### ISSUE TITLE
Native storage issue: natives have only 1 turn worth of materials, making them worthless for trade...

### ISSUE THREAD

**Issue by abmpicoli** [We-the-People-civ4col-mod/Mod#196]
Hi, everyone.

I think that this is some error... It seems that natives now have nothing to offer in trade: they only have one turn worth of raw materials... 

![image](https://user-images.githubusercontent.com/1733170/47250362-2f132700-d3f6-11e8-896b-98464f77d1b5.png)
[natives_with_only-1-turn-worth-of-items.zip](https://github.com/We-the-People-civ4col-mod/Mod/files/2497737/natives_with_only-1-turn-worth-of-items.zip)
Mod named as "We-the-people".

I'm using code from commit ff476c7b3d0b83cbc0acaaed0fa989dcc8b239bb.

**Comment by Nightinggale** [comment]
Considering the amount of change in this branch, I was already starting to wonder if we should postpone merging #75 until after 2.7.1. Now that there seems to be issues, I really think it shouldn't be a 2.7.1 addition. We need to release something stable in the near future.

**Comment by ShadesOT** [comment]
hm, ... i don't see how that commit can leave natives with one turn of production each turn. the changed code never sets the total amount of the yield stored and it did not change how the yields for hurrying were added to the temporary storage (so maybe it could be the exact same number like the yields in the storage but negative). <scratching head>

**Comment by Nightinggale** [comment]
I started looking into this. Inspecting the AI settlements reveal that they do indeed produce those yields, but somehow yields get lost. I suspected CvCityAI::AI_doNative(), but that doesn't seem to be the culprit. I did however examine it enough to open #218, but that won't fix this problem.

**Comment by ShadesOT** [comment]
> ... why is m_bIgnoresBoycott in CvCity rather than CvPlayer?

m_bIgnoresBoycott derives from the properties of buildings, thus it is city specific.

> Why is the AI given a free custom house in all colonies

I have to look that one up. there was a reason for this.

**Comment by ShadesOT** [comment]
@Nightinggale : What you found is part 1 of the bug.
Part 2 of the bug was actually introduced with this commit 11974cf7e9872438d988da8d98fa01fb286e6727 , when I switched the conditions in ```CVCity::doYields()``` 
from
```cpp
// Here special sell behaviour for Custom House
if (getOverflowYieldSellPercent() > 79 && iTotalYields < iMaxCapacity)
```
to
```cpp
// Here special sell behaviour for Custom House
if (bHasUnlockedTradeSettings && iTotalYields < iMaxCapacity)
```
Next: The reason for giving all AI players free
```cpp
m_bIgnoresBoycott = true;
m_bHasUnlockedStorageLossTradeSettings = true;
```		
in each city, is a conversation I had with Erik here #113.
Though at the moment I can not reproduce why I wrote ```Currently it is hardcoded that AI players can sell from their cities warehouses to europe, nonregarding if a good is boycotted.```

Then I did not check, if those replacements I made, integrate correctly with the old customs house code. Because ... it seems, that I have a different understanding about how the customs house works, than Ray had.

And it is dawning on me, that the fundamental cause for all of this, is this very basic misunderstanding or discrepancy of understandings on how the customs house works. 

I am going to write an issue to clarify the customs house mechanics once and for all.

And aside of this chain of incidents and assumptions, to top everything ... the discrepancy of understandings would equate to, that my rewrite of doYields on the refactoring branch now has code for a totaly new game mechanic. Oh boy, ... this now got funny. :-D

**Comment by ShadesOT** [comment]
Issue about clarifying customs house mechanics is here: #219

**Comment by ShadesOT** [comment]
To fix the release we can remove the AI cheat code altogether. As hard as I think I do not know anymore why I gave those cheats to the AI. Maybe just because the AI was and is cheating regarding the overflowSellPercentage. I unbundled this one variable into the 3 variables of
storageLossSellPercentage,
ignoresBoycott
hasUnlockedStorageLossTradeSttings

So maybe I just thought they AI should cheat regarding the new ones as well. When I shift through the code now, I do not see a reason to give these free cheats to the AI europeans.
